// @sentry/nextjs/cjs/server/wrapServerComponentWithSentry.js
Object.defineProperty(exports, '__esModule', { value: true });

const core = require('@sentry/core');
const utils = require('@sentry/utils');
const nextNavigationErrorUtils = require('../common/nextNavigationErrorUtils.js');

/**
 * Wraps an `app` directory server component with Sentry error instrumentation.
 */
function wrapServerComponentWithSentry(
  appDirComponent,
  context,
) {
  core.addTracingExtensions();

  const { componentRoute, componentType } = context;

  // Even though users may define server components as async functions, for the client bundles
  // Next.js will turn them into synchronous functions and it will transform any `await`s into instances of the `use`
  // hook. 🤯
  return new Proxy(appDirComponent, {
    apply: (originalFunction, thisArg, args) => {
      return core.runWithAsyncContext(() => {
        const hub = core.getCurrentHub();

        let maybePromiseResult;

        const traceparentData = context.sentryTraceHeader
          ? utils.extractTraceparentData(context.sentryTraceHeader)
          : undefined;

        const dynamicSamplingContext = utils.baggageHeaderToDynamicSamplingContext(context.baggageHeader);

        const transaction = core.startTransaction({
          op: 'function.nextjs',
          name: `${componentType} Server Component (${componentRoute})`,
          status: 'ok',
          ...traceparentData,
          metadata: {
            source: 'component',
            dynamicSamplingContext: traceparentData && !dynamicSamplingContext ? {} : dynamicSamplingContext,
          },
        });

        const currentScope = hub.getScope();
        if (currentScope) {
          currentScope.setSpan(transaction);
        }

        const handleErrorCase = (e) => {
          if (nextNavigationErrorUtils.isNotFoundNavigationError(e)) {
            // We don't want to report "not-found"s
            transaction.setStatus('not_found');
          } else if (nextNavigationErrorUtils.isRedirectNavigationError(e)) ; else {
            transaction.setStatus('internal_error');
            core.captureException(e);
          }

          transaction.finish();
        };

        try {
          maybePromiseResult = originalFunction.apply(thisArg, args);
        } catch (e) {
          handleErrorCase(e);
          throw e;
        }

        if (typeof maybePromiseResult === 'object' && maybePromiseResult !== null && 'then' in maybePromiseResult) {
          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
          Promise.resolve(maybePromiseResult).then(
            () => {
              transaction.finish();
            },
            e => {
              handleErrorCase(e);
            },
          );

          // It is very important that we return the original promise here, because Next.js attaches various properties
          // to that promise and will throw if they are not on the returned value.
          return maybePromiseResult;
        } else {
          transaction.finish();
          return maybePromiseResult;
        }
      });
    },
  });
}

exports.wrapServerComponentWithSentry = wrapServerComponentWithSentry;
