// @sentry/nextjs/esm/index.server.js
export { ErrorBoundary, IS_BUILD, init, isBuild, showReportDialog, withErrorBoundary } from './server/index.js';
export { withSentryConfig } from './config/withSentryConfig.js';
export { captureUnderscoreErrorException } from './common/_error.js';
export { wrapApiHandlerWithSentryVercelCrons } from './common/wrapApiHandlerWithSentryVercelCrons.js';
export { withSentryGetStaticProps, wrapGetStaticPropsWithSentry } from './server/wrapGetStaticPropsWithSentry.js';
export { withSentryServerSideGetInitialProps, wrapGetInitialPropsWithSentry } from './server/wrapGetInitialPropsWithSentry.js';
export { withSentryServerSideAppGetInitialProps, wrapAppGetInitialPropsWithSentry } from './server/wrapAppGetInitialPropsWithSentry.js';
export { withSentryServerSideDocumentGetInitialProps, wrapDocumentGetInitialPropsWithSentry } from './server/wrapDocumentGetInitialPropsWithSentry.js';
export { withSentryServerSideErrorGetInitialProps, wrapErrorGetInitialPropsWithSentry } from './server/wrapErrorGetInitialPropsWithSentry.js';
export { withSentryGetServerSideProps, wrapGetServerSidePropsWithSentry } from './server/wrapGetServerSidePropsWithSentry.js';
export { withSentry, withSentryAPI, wrapApiHandlerWithSentry } from './server/wrapApiHandlerWithSentry.js';
export { wrapServerComponentWithSentry } from './server/wrapServerComponentWithSentry.js';
export * from '@sentry/node';

// This file is the main entrypoint on the server and/or when the package is `require`d

/**
 * This const serves no purpose besides being an identifier for this file that the SDK multiplexer loader can use to
 * determine that this is in fact a file that wants to be multiplexed.
 */
const _SENTRY_SDK_MULTIPLEXER = true;

export { _SENTRY_SDK_MULTIPLEXER };
